src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/ViewCodeCommandHandler.cs

```diff
@@ -0,0 +1,94 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.ComponentModel.Composition;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using Microsoft.VisualStudio.Commanding;
+using Microsoft.VisualStudio.LanguageServerClient.Razor.VS.LSClientRazor;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor.Commanding.Commands;
+using Microsoft.VisualStudio.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServerClient.Razor
+{
+    [Name(nameof(ViewCodeCommandHandler))]
+    [Export(typeof(ICommandHandler))]
+    [ContentType(RazorLSPConstants.RazorLSPContentTypeName)]
+    internal sealed class ViewCodeCommandHandler : ICommandHandler<ViewCodeCommandArgs>
+    {
+        private const string CSharpFileExtension = ".cs";
+
+        // Because query status happens all the time we want to cache the File.Exists checks for a reasonable amount of time
+        private const int CacheTimeoutMilliseconds = 10000;
+        private static readonly Stopwatch s_fileExistsStopwatch = Stopwatch.StartNew();
+        private static readonly Dictionary<string, (bool exists, long addedMs)> s_fileExistsCache = new();
+
+        private static readonly ImmutableHashSet<string> s_relatedRazorFileSuffixes = ImmutableHashSet.CreateRange(StringComparer.OrdinalIgnoreCase, new[] { ".cshtml", ".razor" });
+
+        private static readonly CommandState s_availableCommandState = new CommandState(isAvailable: true, displayText: Resources.View_Code);
+
+        public string DisplayName => nameof(ViewCodeCommandHandler);
+
+        public CommandState GetCommandState(ViewCodeCommandArgs args)
+        {
+            if (TryGetCSharpFilePath(args.SubjectBuffer, out _))
+            {
+                return s_availableCommandState;
+            }
+
+            return CommandState.Unavailable;
+        }
+
+        public bool ExecuteCommand(ViewCodeCommandArgs args, CommandExecutionContext executionContext)
+        {
+            if (TryGetCSharpFilePath(args.SubjectBuffer, out var cSharpFilePath))
+            {
+                VsShellUtilities.OpenDocument(ServiceProvider.GlobalProvider, cSharpFilePath);
```

Comment:
Hmm, I don't understand this API super well, but it seems like even if we know it exists on disk there are still scenarios we'd have to be defensive about while opening it? Also, does `OpenDocument` automatically focus the new tab?

Reply:
I'm curious what other scenarios you have in mind?

This method focuses the tab yes. Its a helper method that does "the default" so you don't have to use the `IVsUIShellOpenDocument` service directly, which you would use to control more stuff. Source is here if you feel like digging in: http://index/?leftProject=Microsoft.VisualStudio.Shell.15.0&leftSymbol=ylfkfjxq9yys&file=VsShellUtilities.cs&line=626